### PR TITLE
implemented learning rate schedulers in a similar way to current reweighters

### PR DESCRIPTION
**Pull request by theo1701**
There are now two extra arguments:
`scheduler_name` : one of the supported learning rate schedulers from lrscheduler.py. The currently supported options are constant learning rate, linear warming up learning rate (rise from 0 linearly to set learning rate), exponential decay, and linear warm up followed by exponential decay. Nothing happens if provided name is not one of the defined schedulers or left empty (defaults to None).
`reduce_on_plateau` : An integer specifying the number of epoch to wait before reducing learning rate, feed as the patience parameter to ReduceLROnPlateau (https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/ReduceLROnPlateau). It works similarly to the current EarlyStopping callback. The default value is 0 and will be interpreted as not reducing learning rate on plateau, so the keras callback for it will not get created.

In short, some extra supports for fine-tuning learning rates. Not providing the above two arguments, or setting them to `constant` and `0` gives the same behaviour as what currently is.

This branch is still work in progress. The stated functionalities are implemented but they are yet to be fully tested.

Completed Tests just for guaranteeing that it doesn't break things: 
(I'll call the two arguments `sn` and `rop` here)

- [x] closure tests with sn and rop
- [x] linear rw tests with sn and rop
- [x] mc tests with sn and roo

### CODE REVIEW

**Comment by theo1701** [comment]
@zctao

**Comment by theo1701** [comment]
(This is on top of the parallelization branch, for these tests n_models_in_parallel is set to 1)

**Comment by theo1701** [comment]
A known behaviour is that if there is no improvement during warm up, reduce lr on plateau will kick in and reduce the lr. Since the two callbacks are not directly related, there isn't much of a way to avoid this. The effect on lr is that it will reduce quicker than the original expectation.
This can be avoided by setting `reduce_on_plateau` to be greater than `WARM_UP_EPOCHS` in lrscheduler.py. 
Therefore, it wouldn't be a problem. Just something to keep in mind.

**Comment by theo1701** [comment]
Merged changes made in master here

**Comment by theo1701** [comment]
This has been sitting here for a while with all the original planned tests finished. The default behaviour without specifying arguments for lrsscheduler is also identical to without this commit at all. Therefore, this can be merged after reviewing.

File: scripts/unfoldv2.py [hunk 1]
```diff
@@ -252,6 +259,11 @@ def getArgsParser(arguments_list=None, print_help=False):
                         help="If True, run unfolding also with IBU for comparison")
     parser.add_argument('-w', '--weight-type', type=str, default='nominal',
                         help="Type of event weights to retrieve from ntuples")
+    parser.add_argument('--scheduler-names', help="Name of the learning rate scheduler to be used", type=str, nargs='+')
+    parser.add_argument('--reduce-on-plateau', type=int, default=0,
+                        help="number of epoch to wait before reducing learning rate")
+    parser.add_argument('--scheduler-args', help="extra arguments for the requested scheduler", type=str)
+    parser.add_argument('--learning-rate', type=float, default=0.001, help="Initial learning rate")
```
Comment (zctao) [inline scripts/unfoldv2.py:266]: Would it be possible to group all these arguments into one e.g. `--scheduler-config`, which takes a config file? And the default file would contain the best config that you recommend.

File: python/modelUtils.py [hunk 2]
```diff
@@ -47,10 +56,9 @@ def get_callbacks(model_filepath=None):
 
         logger_fp = model_filepath + "_history.csv"
         CSVLogger = keras.callbacks.CSVLogger(filename=logger_fp, append=False)
-
-        return [CheckPoint, CSVLogger, EarlyStopping]
+        return [cb for cb in [CheckPoint, CSVLogger, EarlyStopping, Reduce] if cb is not None] + lr_callbacks + custom_callbacks
```
Comment (zctao) [inline python/modelUtils.py:59]: Hmm this looks unnecessarily convoluted (same for the other one below). How about just having a `callbacks = []`, and add to the list with if statement? The only cb that could be None is `Reduce`.

Comment (zctao) [reply python/modelUtils.py:59]: And can `Reduce` and `PrintLearningRate` be configured elsewhere and included together with other ones in `lr_callbacks` for better encapsulation?

File: python/OmniFoldTTbar.py [hunk 3]
```diff
@@ -2,6 +2,7 @@
 import glob
 import numpy as np
 from copy import copy
+from python.lrscheduler import constant
```
[the remark this answers is not in the document]
Comment (theo1701) [reply python/OmniFoldTTbar.py:5]: `constant` isn't used in OmniFoldTTbar and it doesn't make sense to import a learning rate function directly, will be removed.

File: python/modelUtils.py [hunk 4]
```diff
@@ -34,6 +38,11 @@ def get_callbacks(model_filepath=None):
         monitor="val_loss", patience=10, verbose=1, restore_best_weights=True
     )
 
+    Reduce = keras.callbacks.ReduceLROnPlateau(monitor='val_loss', factor = 0.2, patience = reduce_on_plateau) if reduce_on_plateau != 0 else None
+    custom_callbacks = [PrintLearningRate()]
```
[the remark this answers is not in the document]
Comment (theo1701) [reply python/modelUtils.py:42]: I find it nice to see current learning rate together with the loss and more assuring. The `PrintLearningRate` callback is also using log at debug level.

File: scripts/unfoldv2.py [hunk 5]
```diff
@@ -92,6 +94,10 @@ def unfold(**parsed_args):
         data_reweighter = rw,
         weight_type = parsed_args["weight_type"]
         )
+    
+    scheduler_args = json.loads(parsed_args["scheduler_args"]) if parsed_args["scheduler_args"] is not None else None
+
+    init_lr_scheduler(parsed_args["learning_rate"], parsed_args["scheduler_names"], scheduler_args)
```
Comment (theo1701) [inline scripts/unfoldv2.py:100]: > One general comment: it would be great if we could better encapsulate the learning rate & scheduler stuff. For example, I'm not sure about the `reduce_on_plateau` argument passed in multiple levels at many places. Would it help if we wrap all relevant stuff into a `LRScheduler` (or other better names), and make it a member of `OmniFoldTTbar`? Or make it global?

Good point, it is pretty tedious. I'll combine the options into a single config file and handle all initialization inside lrscheduler class directly. (the same approach as `preprocessor`)

**Comment by theo1701** [comment]
- [ ] linear reweigh test

**Comment by theo1701** [comment]
- [ ] mc data test with 50 runs on default option

File: configs/run/quick_run.json [hunk 6]
```diff
@@ -2,7 +2,8 @@
   "data" : ["/data/ztao/ttbarDiffXs13TeV/latest/mcWAlt/ttbar_nominal/mc16a/ttbar_0_pseudotop_parton_ejets.root"],
   "signal" : ["/data/ztao/ttbarDiffXs13TeV/latest/mcWAlt/ttbar_nominal/mc16a/ttbar_1_pseudotop_parton_ejets.root"],
   "observable_config" : "configs/observables/vars_ttbardiffXs_pseudotop.json",
-  "iterations" : 3,
+  "iterations" : 10,
+  "nruns":1,
```
Comment (zctao) [inline configs/run/quick_run.json:6]: I guess this file is committed by accident instead of the default config for `--lrscheduler-config`: `configs/lrs/constant_warm_up.json`?

File: scripts/unfoldv2.py [hunk 7]
```diff
@@ -5,6 +5,7 @@
 import tracemalloc
 import logging
 import numpy as np
+import json
```
Comment (zctao) [inline scripts/unfoldv2.py:8]: No need to import json here

**Review by zctao — COMMENTED** [review]
Thanks for the updates! Looks good. 
If you can also put a few example config files in `configs/lrs/` that would be awesome.

File: python/modelUtils.py [hunk 8]
```diff
@@ -23,6 +24,8 @@ def get_callbacks(model_filepath=None):
 
     Parameters
     ----------
+    reduce_on_plateau : int, optional
+        If not 0, the model will reduce learning rate by a certain multiplicative factor after set epoch without improvement
```
Comment (zctao) [inline python/modelUtils.py:28]: `reduce_on_plateau` is no longer a parameter for `get_callbacks`

File: python/lrscheduler.py [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+# note two methods can not be used together, since using schedules takes up the place for learning rate
+
+import numpy as np
+import tensorflow.keras.optimizers.schedules as schedules
+import tensorflow.keras.callbacks as callbacks
+import json
+from callbacks import PrintLearningRate
+
+WARM_UP_EPOCHS = 5
+PLATEAU = 10
```
Comment (zctao) [inline python/lrscheduler.py:15]: Is `PLATEAU` actually used somewhere?

File: python/lrscheduler.py [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+# The two methods each has some nice features that are not trivially available in the other
+# note two methods can not be used together, since using schedules takes up the place for learning rate
+
+import numpy as np
+import tensorflow.keras.optimizers.schedules as schedules
+import tensorflow.keras.callbacks as callbacks
+import json
+from callbacks import PrintLearningRate
+
+WARM_UP_EPOCHS = 5
```
Comment (zctao) [inline python/lrscheduler.py:14]: I'm not sure about this but would it be better to have `WARM_UP_EPOCHS` in the config?

File: python/modelUtils.py [hunk 11]
```diff
@@ -9,10 +9,11 @@
 from tensorflow.keras import layers
 import tensorflow.keras.backend as K
 from sklearn.model_selection import train_test_split
+from lrscheduler import get_lr_scheduler
 
 import plotter
 
-n_models_in_parallel = 1
+n_models_in_parallel = 5
```
Comment (zctao) [inline python/modelUtils.py:16]: Let's keep the default to 1.

File: python/lrscheduler.py (hunk 9 above)
Comment (theo1701) [reply python/lrscheduler.py:15]: It was for one of the removed schedules. Removed now.

File: python/modelUtils.py (hunk 11 above)
Comment (theo1701) [reply python/modelUtils.py:16]: thanks for pointing this out. I just keep forgetting about this

Comment (theo1701) [reply python/modelUtils.py]: At the moment there is only this one setting for model configs. I'll note this down and implement it later. It will be convenient to write custom model structures directly in model_config

**Comment by theo1701** [comment]
Now the schedules are more customizable, with each accepting extra arguments. There are currently two lr configs. One is used as the default best option. Behind the lrscheduler, one of them is a callback and the other is a scheduler. So it should provide good template for writing more configs.

**Comment by zctao** [comment]
Great. Thanks!

**Comment by theo1701** [comment]
Added two configs. Constant + rlop and inverse time decay sometimes performs better than the default constant + warmup, but that is not guaranteed. On the other hand, the default option is tested to consistently perform better than without any learning rate schedule (just constant learning rate)

**Comment by theo1701** [comment]
Everything is now tested with linear reweigh. The schedules themselves were tested on mc data previously. There shouldn't be any correlation between implementation and the type of data used.
